### PR TITLE
[MM-47090] webapp/channels : Migrate "components/integrations/abstract_command.jsx" to Typescript

### PR DESCRIPTION
**Pull request by saideepesh000**
#### Summary
migrate `abstract_command.jsx` to typescript

#### Ticket Link

fixes https://github.com/mattermost/mattermost-server/issues/21079
https://mattermost.atlassian.net/browse/MM-47090

#### Release Note

```release-note
NONE
```

### CODE REVIEW

File: webapp/channels/src/components/integrations/abstract_command.tsx [hunk 1]
```diff
@@ -385,15 +379,15 @@ export default class AbstractCommand extends React.PureComponent {
...
+                                    maxLength={128}
                                     className='form-control'
                                     value={this.state.description}
                                     onChange={this.updateDescription}
                                 />
                                 <div className='form__help'>
                                     <FormattedMessage
                                         id='add_command.description.help'
-                                        defaultMessage='Describe your slash command.'
+                                        defaultMessage='Describe your incoming webhook.'
```
Comment (sinansonmez) [inline webapp/channels/src/components/integrations/abstract_command.tsx:390]: Why do we need to change the default message from slash command to incoming webhook?

File: webapp/channels/src/components/integrations/abstract_command.tsx [hunk 2]
```diff
@@ -436,15 +430,16 @@ export default class AbstractCommand extends React.PureComponent {
                                         defaultMessage='Reserved: {link}'
                                         values={{
                                             link: (
-                                                <ExternalLink
+                                                <a
                                                     href='https://developers.mattermost.com/integrate/admin-guide/admin-slash-commands/#built-in-commands'
-                                                    location='abstract_command'
+                                                    target='_blank'
```
Comment (sinansonmez) [inline webapp/channels/src/components/integrations/abstract_command.tsx:435]: style checker is complaining about it.

File: webapp/channels/src/components/integrations/abstract_command.tsx [hunk 3]
```diff
@@ -436,15 +430,16 @@ export default class AbstractCommand extends React.PureComponent {
                                         defaultMessage='Reserved: {link}'
                                         values={{
                                             link: (
-                                                <ExternalLink
+                                                <a
```
Comment (sinansonmez) [inline webapp/channels/src/components/integrations/abstract_command.tsx:433]: Why are we changing `ExternalLink` to `a`?

**Review by sinansonmez — CHANGES_REQUESTED** [review]
Thanks @saideepesh000 for the great work. I just have a few comments and questions.

**Comment by saideepesh000** [comment]
> Thanks @saideepesh000 for the great work. I just have a few comments and questions.

thanks @sinansonmez , made the changes

File: webapp/channels/src/components/integrations/abstract_command.test.tsx [hunk 4]
```diff
@@ -47,14 +47,13 @@ describe('components/integrations/AbstractCommand', () => {
         header,
         footer,
         loading,
-        renderExtra: 'renderExtra',
```
[the remark this answers is not in the document]
Comment (sinansonmez) [reply webapp/channels/src/components/integrations/abstract_command.test.tsx:50]: Resolved

File: webapp/channels/src/components/integrations/abstract_command.test.tsx [hunk 5]
```diff
@@ -109,67 +108,78 @@ describe('components/integrations/AbstractCommand', () => {
...
         const props = {...baseProps, action: newAction};
-        const wrapper = shallow(
+        const wrapper = shallow<AbstractCommand>(
             <AbstractCommand {...props}/>,
         );
         expect(newAction).toHaveBeenCalledTimes(0);
 
-        const evt = {preventDefault: jest.fn()};
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        const evt: any = {preventDefault: jest.fn()};
```
[the remark this answers is not in the document]
Comment (sinansonmez) [reply webapp/channels/src/components/integrations/abstract_command.test.tsx]: Resolved

File: webapp/channels/src/components/integrations/abstract_command.tsx [hunk 6]
```diff
@@ -5,29 +5,30 @@ import React, {ChangeEvent} from 'react';
...
 type Props = {
-    team: Team;
-    header: {id: string ; defaultMessage: string};
-    footer: {id: string ; defaultMessage: string};
-    loading: {id: string ; defaultMessage: string};
-    renderExtra?: JSX.Element;
-    serverError: string;
-    initialCommand?: Partial<Command>;
-    action: (command: Command) => Promise<void>;
+    team: Team; // The current team
```
Comment (sinansonmez) [inline webapp/channels/src/components/integrations/abstract_command.tsx:24]: @M-ZubairAhmed do we need to keep also multiline commenting for the the tool for documentation or single line comments are also fine?

File: webapp/channels/src/components/integrations/abstract_command.tsx [hunk 7]
```diff
@@ -14,79 +14,62 @@ import SpinnerButton from 'components/spinner_button';
...
-        /**
-        * The async function to run when the action button is pressed
-        */
-        action: PropTypes.func.isRequired,
-    };
+type Props = {
+    team: Team;
+    header: {id: string ; defaultMessage: string};
+    footer: {id: string ; defaultMessage: string};
+    loading: {id: string ; defaultMessage: string};
```
[the remark this answers is not in the document]
Comment (sinansonmez) [reply webapp/channels/src/components/integrations/abstract_command.tsx]: Resolved

File: webapp/channels/src/components/integrations/abstract_command.tsx (hunk 3 above)
Comment (sinansonmez) [reply webapp/channels/src/components/integrations/abstract_command.tsx]: Rosolved

**Review by sinansonmez — COMMENTED** [review]
Thanks for the changes @saideepesh000 just one last point left and one question for Zubair?

File: webapp/channels/src/components/integrations/abstract_command.tsx (hunk 2 above)
Comment (saideepesh000) [reply webapp/channels/src/components/integrations/abstract_command.tsx:436]: oh sorry, didn't check this, I was testing something

Comment (saideepesh000) [reply webapp/channels/src/components/integrations/abstract_command.tsx]: made the changes

File: webapp/channels/src/components/integrations/abstract_command.tsx [hunk 8]
```diff
@@ -12,81 +11,66 @@ import * as Utils from 'utils/utils';
...
-
-        /**
-        * The async function to run when the action button is pressed
-        */
-        action: PropTypes.func.isRequired,
-    };
+type Props = {
+    team: Team; // The current team
+    header: {id: string; defaultMessage: string}; // The header text to render, has id and defaultMessage
+    footer: {id: string; defaultMessage: string}; // The footer text to render, has id and defaultMessage
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/integrations/abstract_command.tsx:26]: Can this be MessageDescriptor from react-intl itself.

File: webapp/channels/src/components/integrations/abstract_command.tsx [hunk 9]
```diff
@@ -113,11 +97,19 @@ export default class AbstractCommand extends React.PureComponent {
...
             auto_complete: this.state.autocomplete,
             team_id: this.props.team.id,
+            auto_complete_desc: '',
+            auto_complete_hint: '',
+            token: '',
+            create_at: 0,
+            update_at: 0,
+            delete_at: 0,
+            id: '',
+            creator_id: '',
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/integrations/abstract_command.tsx:107]: why did we had to reset the values here?

File: webapp/channels/src/components/integrations/add_command/add_command.tsx [hunk 10]
```diff
@@ -58,7 +58,6 @@ const AddCommand = ({team, actions}: Props) => {
             header={HEADER}
             footer={FOOTER}
             loading={LOADING}
-            renderExtra={''}
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/integrations/add_command/add_command.tsx:61]: Was this unused property?

File: webapp/channels/src/components/integrations/add_command/add_command.tsx [hunk 11]
```diff
@@ -58,7 +58,6 @@ const AddCommand = ({team, actions}: Props) => {
             header={HEADER}
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/integrations/add_command/add_command.tsx:58]: Although this is unrelated to this PR, is it possible to pass the formatmessage directly here instead of a 't' function?
Eg.
```
header={formatMessage({id: 'intergration.ad............
```

File: webapp/channels/src/components/integrations/abstract_command.tsx (hunk 6 above)
Comment (saideepesh000) [reply webapp/channels/src/components/integrations/abstract_command.tsx:24]: Sure, will make the changes

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]
Looks good to me, with a few suggestions.

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]
Looks good to me, with a few suggestions.

File: webapp/channels/src/components/integrations/abstract_command.tsx (hunk 8 above)
Comment (saideepesh000) [reply webapp/channels/src/components/integrations/abstract_command.tsx:26]: yes, makes sense

File: webapp/channels/src/components/integrations/abstract_command.tsx (hunk 9 above)
Comment (saideepesh000) [reply webapp/channels/src/components/integrations/abstract_command.tsx:107]: We're initializing the object here so that we don't get typescript error like this
Screenshot 2023-05-08 at 5 49 22 PM

File: webapp/channels/src/components/integrations/add_command/add_command.tsx (hunk 10 above)
Comment (saideepesh000) [reply webapp/channels/src/components/integrations/add_command/add_command.tsx:61]: so, instead of passing the empty string, I just removed as it is of type JSX element

File: webapp/channels/src/components/integrations/add_command/add_command.tsx (hunk 11 above)
Comment (saideepesh000) [reply webapp/channels/src/components/integrations/add_command/add_command.tsx:58]: So t function is for translation here, and since it is an object and has 3 variants like this, it's better to have them separately.

Comment (saideepesh000) [reply webapp/channels/src/components/integrations/add_command/add_command.tsx:58]: Screenshot 2023-05-08 at 5 58 19 PM

**Review by sinansonmez — APPROVED** [review]
LGTM 🚀

File: webapp/channels/src/components/integrations/add_command/add_command.tsx [hunk 12]
```diff
@@ -35,7 +30,10 @@ export type Props = {
 
 const AddCommand = ({team, actions}: Props) => {
     const history = useHistory();
-
+    const {formatMessage} = useIntl();
+    const HEADER = formatMessage({id: ('integrations.add'), defaultMessage: 'Add'}) as MessageDescriptor;
+    const FOOTER = formatMessage({id: ('add_command.save'), defaultMessage: 'Save'}) as MessageDescriptor;
+    const LOADING = formatMessage({id: ('add_command.saving'), defaultMessage: 'Saving...'}) as MessageDescriptor;
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/integrations/add_command/add_command.tsx:36]:
```suggestion
    const headerMessage = formatMessage({id: ('integrations.add'), defaultMessage: 'Add'}) as MessageDescriptor;
    const footerMessage = formatMessage({id: ('add_command.save'), defaultMessage: 'Save'}) as MessageDescriptor;
    const loadingMessage = formatMessage({id: ('add_command.saving'), defaultMessage: 'Saving...'}) as MessageDescriptor;
```

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]
There are some failing snaphots, can you please update them
```
npm run test -- -u
```

**Review by M-ZubairAhmed — APPROVED** [review]
This looks good :shipit: Thank you

**Comment by saideepesh000** [comment]
@M-ZubairAhmed , test failed again

**Comment by M-ZubairAhmed** [comment]
@saideepesh000 it failed again :P

**Comment by saideepesh000** [comment]
> @saideepesh000 it failed again :P

There isn't anything wrong happened, it just failed randomly

**Comment by saideepesh000** [comment]
@M-ZubairAhmed , found the issue, there was an obselete jsx snap file, I deleted it, please run the tests again

**Review by yasserfaraazkhan — APPROVED** [review]

**Comment by M-ZubairAhmed** [comment]
@yasserfaraazkhan is this good to merge i do see a few E2E failing though. 🤔

**Comment by saideepesh000** [comment]
@M-ZubairAhmed @yasserfaraazkhan , can anyone check this

**Comment by M-ZubairAhmed** [comment]
@saideepesh000 apologies for holding this up :)

**Comment by M-ZubairAhmed** [comment]
@yasserfaraazkhan can you please help verify if the failed test would be flaky for latest release

**Comment by yasserfaraazkhan** [comment]
@M-ZubairAhmed not seeing anything in cycle, will re-trigger
Screenshot 2023-06-14 at 1 54 34 AM

**Comment by M-ZubairAhmed** [comment]
@yasserfaraazkhan it failed again. re triggered.
